Support auxiliary files in terraform deploy spec

endorama opened the pull request:
## What does this PR do?

<!-- Mandatory
Explain here WHAT changes you made in the PR.
-->

- add support for Terraform dependency lock file
- add support for Terraform template files
- add support for JSON files

## Why is it important?

<!-- Mandatory
Explain here the WHY, or the rationale/motivation for the changes.
-->

- Terraform suggest to add it's dependency lock file to VCS, as it helps with reproducible installations
- Terraform support template files, with `.tftpl` suggested extension
- JSON files may be used for ausiliary content required by Terraform code.

Further discussion available in #269.

## Checklist

<!-- Mandatory
Add a checklist of things that are required to be reviewed in order to have the PR approved

List here all the items you have verified BEFORE sending this PR. Please DO NOT remove any item, striking through those that do not apply. (Just in case, strikethrough uses two tildes. ~~Scratch this.~~)
-->

- [x] I have added test packages to [`test/packages`](https://github.com/elastic/package-spec/tree/main/test/packages) that prove my change is effective.
- [x] I have added an entry in [`versions/N/changelog.yml`](https://github.com/elastic/package-spec/blob/main/versions/1/changelog.yml).

## Related issues

<!-- Recommended
Link related issues below. Insert the issue link or reference after the word "Closes" if merging this should automatically close it.

- Closes #123
- Relates #123
- Requires #123
- Supersedes #123
-->
- Closes #269

endorama commented:
@mtojek I have a question about the changelog. The file header says 
```
[...]
## Newer entries go at the bottom.
[...]
```

but it seems the opposite is true? Am I expected to append to `v1.4.1` as is still unreleased or create a new version?

elasticmachine commented:
## :broken_heart: Build Failed


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fpackage-spec%2FPR-272/detail/PR-272/2//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fpackage-spec%2FPR-272/detail/PR-272/2//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fpackage-spec%2FPR-272/detail/PR-272/2//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fpackage-spec%2FPR-272/detail/PR-272/2//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://package-spec_272.docs-preview.app.elstc.co/diff)  [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/beats-ci/transactions/view?rangeFrom=2022-02-09T09:56:11.135Z&rangeTo=2022-02-09T10:16:11.135Z&transactionName=Ingest-manager/package-spec/PR-{number}&transactionType=job&latencyAggregationType=avg&traceId=8146fedd6f296f1193eb7daf2095a2c2&transactionId=9a35048e37320ebe)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats




* Start Time: 2022-02-09T10:06:11.135+0000


* Duration: 5 min 49 sec





</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->











### Steps errors [![1](https://img.shields.io/badge/1%20-red)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fpackage-spec%2FPR-272/detail/PR-272/2//pipeline)
  <details><summary>Expand to view the steps failures</summary>
  <p>


  
  
  
  
##### `Test`
<ul>
<li>Took 0 min 5 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Ingest-manager/pipelines/package-spec/pipelines/PR-272/runs/2/steps/195/log/?start=0">here</a></li>
<li>Description: <code>make test</code></l1>
</ul>
  
  </p>
  </details>


## :robot: GitHub comments

To re-run your PR in the CI, just comment with:



- `/test` : Re-trigger the build.




<!--COMMENT_GENERATED_WITH_ID_comment.id-->

jsoriano commented:
> but it seems the opposite is true? Am I expected to append to `v1.4.1` as is still unreleased or create a new version?

Put it at the bottom of the 1.4.1 section, yes :)

jsoriano commented on versions/1/changelog.yml:
```diff
@@ -10,6 +10,9 @@
     link: https://github.com/elastic/package-spec/pull/265
+  - description: Support ausiliary files in terraform deploy spec
+    type: enhancement
+    link: https://github.com/elastic/package-spec/pull/272
```
:+1:

endorama commented:
I added a `deploy_terraform` test package, but I'm not sure is being used :thinking: given that I also added the same files to package `good`, should I remove it?

jsoriano commented:
> I added a `deploy_terraform` test package, but I'm not sure is being used thinking given that I also added the same files to package `good`, should I remove it?

Test packages are used in `code/go/pkg/validator/validator_test.go`, you will need to add the package there.

jsoriano commented on test/packages/deploy_terraform/_dev/deploy/docker/docker-compose.yml:
```diff
@@ -0,0 +1,3 @@
+version: '2.3'
+services:
+  hello_world:
```
Nit. Is this file needed here? I guess it doesn't do any harm though.

endorama replied:
leftover from copying over the `deploy_docker` module, thank you for catching this.

jsoriano commented on test/packages/good/data_stream/foo/_dev/deploy/tf/env.yml:
```diff
@@ -2,8 +2,5 @@ version: '2.3'
-      - AWS_PROFILE=${AWS_PROFILE}
-      - AWS_REGION=${AWS_REGION:-us-east-1}
\ No newline at end of file
+      - FOO=${BAR:-default-value}
```
Unrelated changes?

endorama replied:
No but given the file is not required I'll remove it

jsoriano commented on versions/1/changelog.yml:
```diff
@@ -10,6 +10,9 @@
   - description: Prepare for next version
     type: enhancement
     link: https://github.com/elastic/package-spec/pull/265
+  - description: Support ausiliary files in terraform deploy spec
```
Nit. Should it be "auxiliary"?
```suggestion
  - description: Support auxiliary files in terraform deploy spec
```

jsoriano commented on versions/1/_dev/deploy/tf/spec.yml:
```diff
@@ -5,9 +5,22 @@ spec:
+      pattern: '^*.json$'
+      type: file
+      contentMediaType: "application/json"
+      required: false
```
Perhaps we end-up needing `additionalContents: false` if we find scenarios that would need arbitrary files. But we can reduce restrictions in the future if we find these cases.

jsoriano reviewed (dismissed):
Some nitpicking, but LGTM in general.

jsoriano commented on versions/1/_dev/deploy/tf/spec.yml:
```diff
@@ -5,9 +5,22 @@ spec:
+    - description: Terraform dependency lock file
+      pattern: ".terraform.lock.hcl"
+      type: file
+      required: true
```
Btw, this may be breaking for current packages. At least we will need to add this file to the AWS package when updating the spec there.
https://github.com/elastic/integrations/tree/8e600daead4b89e97271e4976b984722c8db7c74/packages/aws/data_stream/ec2_metrics/_dev/deploy/tf

endorama replied:
Nice find, actually I don't want to introduce a breaking change here. Do we have a process here? I will make this not required, but I think it make sense to have it everywhere. Is there a way to notify users of this change? For example outputting a message: "this field is going to default to true from version x.y.z"

jsoriano replied:
Maybe we can add this as optional by now, add after some time make it required. At the moment I think that only the AWS package uses it, so not many users affected :)

mtojek replied:
Yeah, let's mark it optional for now.

mtojek reviewed (approved):
LGTM
